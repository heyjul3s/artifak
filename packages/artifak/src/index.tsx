export {
  createStyledComponent,
  createBaseComponents
} from '@artifak/component-generator';

export { BlockBase, createBlockComponents } from '@artifak/block';
export { BlockBaseProps } from '@artifak/block';

export { FlexRow, FlexRowBase, FlexCol, FlexColBase } from '@artifak/flex';
export { FlexRowBaseProps, FlexColBaseProps } from '@artifak/flex';

export { Grid, GridBase, GridItem, GridItemBase } from '@artifak/grid';
export { GridBaseProps, GridItemBaseProps } from '@artifak/grid';

export {
  createTypographyComponents,
  fluidSizing,
  fontWeight,
  TypographyBase
} from '@artifak/typography';
export { TypographyBaseProps } from '@artifak/typography';

export { Imagery, ImageryBase, ImageryProps } from '@artifak/imagery';

export { media, Media } from '@artifak/media';
