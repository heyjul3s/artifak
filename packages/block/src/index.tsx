export { createBlockComponents } from './createBlockComponents';
export { BlockBase } from './BlockBase';
export { BlockBaseProps } from './typings';
