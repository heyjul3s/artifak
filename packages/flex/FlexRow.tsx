import React, { forwardRef } from 'react';
import { FlexRowBase } from './FlexBase';

export const FlexRow = forwardRef((props: any, ref) => {
  return (
    <FlexRowBase
      flexDirection={setFlexDirection(props.reverse)}
      ref={ref}
      {...props}
    >
      {props.children}
    </FlexRowBase>
  );
});

export function setFlexDirection(isReverse: boolean) {
  return !isReverse ? 'row' : 'row-reverse';
}
