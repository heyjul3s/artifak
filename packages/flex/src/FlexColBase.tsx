import { flexbox } from 'styled-system';
import { createStyledComponent } from '@artifak/component-generator';
import { FlexColBaseProps } from './typings';

export const FlexColBase = createStyledComponent<FlexColBaseProps>({
  styles: {
    boxSizing: 'border-box',
    position: 'relative',
    flex: '0 0 auto'
  },
  styleProps: [flexbox]
});
