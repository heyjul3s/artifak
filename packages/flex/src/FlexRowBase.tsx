import { flexbox } from 'styled-system';
import { createBaseStyledComponent } from '@artifak/component-generator';
import { FlexRowBaseProps } from './typings';

export const FlexRowBase = createBaseStyledComponent<FlexRowBaseProps>(
  [flexbox],
  {
    display: 'flex',
    flex: '0 1 auto',
    flexWrap: 'wrap'
  }
);
