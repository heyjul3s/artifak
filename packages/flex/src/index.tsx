export { FlexRow } from './FlexRow';
export { FlexRowBase } from './FlexRowBase';
export { FlexCol } from './FlexCol';
export { FlexColBase } from './FlexColBase';
export { FlexColBaseProps, FlexRowBaseProps } from './typings';
