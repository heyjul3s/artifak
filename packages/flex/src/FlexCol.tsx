import React from 'react';
import { FlexColBase } from './FlexColBase';

/* eslint-disable @typescript-eslint/no-explicit-any */
export const FlexCol = React.forwardRef((props: any, ref) => {
  const baseColumnWidth = getColumnWidth(props.totalColumns);
  const columnOffset = setColumnOffset(baseColumnWidth, props.offset);
  const columnSizing = setColumnSizing(baseColumnWidth, props.columnSize);
  const gutterWidth = setGutterWidth(props.gutterWidth);

  return (
    <FlexColBase
      ml={columnOffset}
      maxWidth={columnSizing}
      flexBasis={columnSizing}
      pl={gutterWidth}
      pr={gutterWidth}
      style={{
        boxSizing: 'border-box'
      }}
      ref={ref}
      {...props}
    >
      {props.children}
    </FlexColBase>
  );
});

export function getColumnWidth(totalColumns = 12): number {
  return isNumber(totalColumns) && totalColumns > 2
    ? Number(parseFloat((100 / totalColumns).toString()).toFixed(2))
    : 100;
}

export function setColumnSizing(
  baseColumnWidth: number,
  columnSize: number | number[]
): string | string[] {
  return Array.isArray(columnSize) && columnSize.length > 1
    ? columnSize.map(
        size => `${parseNumberToTwoDecimals(size * baseColumnWidth)}%`
      )
    : `${parseNumberToTwoDecimals((columnSize as number) * baseColumnWidth)}%`;
}

export function setColumnOffset(
  baseColumnWidth: number,
  offset: number | number[]
): string | string[] | undefined {
  if (Array.isArray(offset) && offset.length > 1) {
    return offset.map(size => `${size * baseColumnWidth}%`);
  }

  if (!Array.isArray(offset) && !!offset) {
    return `${(offset as number) * baseColumnWidth}%`;
  }

  return void 0;
}

export function setGutterWidth(
  gutterWidth: number | number[] = 1
): string | string[] {
  return Array.isArray(gutterWidth) && gutterWidth.length > 1
    ? gutterWidth.map(width => `${width / 2}em`)
    : `${(gutterWidth as number) / 2}em`;
}

export function parseNumberToTwoDecimals(value: number): string {
  return `${parseFloat(String(value)).toFixed(2)}`;
}

export function isNumber(val: number): boolean {
  return !isNaN(val) && isFinite(val);
}
