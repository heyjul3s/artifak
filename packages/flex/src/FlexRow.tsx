import React from 'react';
import { FlexRowBase } from './FlexRowBase';
import { Property } from 'csstype';

/* eslint-disable @typescript-eslint/no-explicit-any */
export const FlexRow = React.forwardRef((props: any, ref) => {
  const flexDirection = setFlexDirection(props.reverse);

  return (
    <FlexRowBase flexDirection={flexDirection} ref={ref} {...props}>
      {props.children}
    </FlexRowBase>
  );
});

export function setFlexDirection(isReverse: boolean): Property.FlexDirection {
  return !isReverse ? 'row' : 'row-reverse';
}
