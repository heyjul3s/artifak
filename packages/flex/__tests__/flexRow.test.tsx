import React from 'react';
import { render } from '@testing-library/react';
import { FlexRow, setFlexDirection } from '../src/FlexRow';

describe('FlexRow', () => {
  describe('setFlexDirection', () => {
    it('should return flex-direction of "row" if passed a boolean value of FALSE', () => {
      const expected = 'row';
      expect(setFlexDirection(false)).toEqual(expected);
    });

    it('should return flex-direction of "row-reverse" if passed a boolean value of TRUE', () => {
      const expected = 'row-reverse';
      expect(setFlexDirection(true)).toEqual(expected);
    });
  });

  describe('FlexRow component', () => {
    it('should render', () => {
      const { asFragment } = render(<FlexRow>Flex Row</FlexRow>);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render row-reverse', () => {
      const { asFragment } = render(
        <FlexRow reverse={true}>Flex Row Reverse</FlexRow>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
