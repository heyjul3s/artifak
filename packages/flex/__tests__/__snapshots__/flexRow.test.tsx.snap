// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexRow FlexRow component should render 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ jnIjoz"
    display="flex"
  >
    Flex Row
  </div>
</DocumentFragment>
`;

exports[`FlexRow FlexRow component should render row-reverse 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ bumAni"
    display="flex"
  >
    Flex Row Reverse
  </div>
</DocumentFragment>
`;
